src/aws-cpp-sdk-core/include/aws/core/client/UserAgent.h

```diff
@@ -0,0 +1,45 @@
+/**
+ * Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
+ * SPDX-License-Identifier: Apache-2.0.
+ */
+#pragma once
+#include <aws/core/client/ClientConfiguration.h>
+#include <aws/core/client/RetryStrategy.h>
+#include <aws/core/utils/memory/stl/AWSSet.h>
+
+namespace Aws {
+namespace Client {
+
+enum class UserAgentFeature {
+  RETRY_MODE_LEGACY,
+  RETRY_MODE_STANDARD,
+  RETRY_MODE_ADAPTIVE,
+  S3_TRANSFER,
+  S3_CRYPTO_V1N,
+  S3_CRYPTO_V2,
+};
+
+class AWS_CORE_API UserAgent {
+ public:
+  explicit UserAgent(const ClientConfiguration& clientConfiguration, std::shared_ptr<Aws::Client::RetryStrategy> retryStrategy);
+  Aws::String SerializeWithFeatures(const Aws::Set<UserAgentFeature>& features) const;
+  void SetApiName(const Aws::String& apiName) { m_api = apiName; }
+  void AddLegacyFeature(const Aws::String& legacyFeature);
+
+ private:
+  Aws::String m_sdkVersion;
+  Aws::String m_userAgentVersion;
+  Aws::String m_api;
+  Aws::String m_crtVersion;
+  Aws::String m_osVersion;
+  Aws::String m_archName;
+  Aws::String m_cppVersion;
+  Aws::String m_compilerMetadata;
+  std::shared_ptr<Aws::Client::RetryStrategy> m_retryStrategy;
```

Comment:
I'm not really sure we need a whole shared pointer the the actual retryStrategy used
It would make sense if we were also to report some retry metrics, but it appears we only report strategy name.
It is a kinda nitpick to think about.

Reply:
so the reason i do this is because `GetStrategyName` is [declared virtual](https://github.com/aws/aws-sdk-cpp/blob/main/src/aws-cpp-sdk-core/include/aws/core/client/RetryStrategy.h#L81-L85) and i was trying to stick by the "[Dont call virtual methods in constructors](https://wiki.sei.cmu.edu/confluence/display/cplusplus/OOP50-CPP.+Do+not+invoke+virtual+functions+from+constructors+or+destructors)" guidance so i defer the call to when we actually assemble the User Agent. the retry strategy should be well defined when we call it in the AwsClient constructor, but i just wanted to avoid it all together.

I actaully had it the other way to begin with and when back on it, could go either way, lemme know if you prefer one way or the other.

Reply:
talked offline -- we only need to follow this when its a virtual method of the class itself we're constructing, todo - move back to using name